src/big/big_decimal.cr

```diff
@@ -47,39 +47,78 @@ struct BigDecimal < Number
   def initialize(str : String)
     raise InvalidBigDecimalException.new(str, "Zero size") if str.bytesize == 0
 
-    # Check str's validity and find index of .
+    # Check str's validity and find index of '.'
     decimal_index = nil
+    # Check str's validity and find index of 'e'
+    exponent_index = nil
+
     str.each_char_with_index do |char, index|
       case char
       when '-'
-        if index != 0
+        unless index == 0 || exponent_index == index - 1
           raise InvalidBigDecimalException.new(str, "Unexpected '-' character")
         end
+      when '+'
+        unless exponent_index == index - 1
+          raise InvalidBigDecimalException.new(str, "Unexpected '+' character")
+        end
       when '.'
         if decimal_index
           raise InvalidBigDecimalException.new(str, "Unexpected '.' character")
         end
-
         decimal_index = index
+      when 'e', 'E'
+        if exponent_index
+          raise InvalidBigDecimalException.new(str, "Unexpected #{char.inspect} character")
+        end
+        exponent_index = index
       when '0'..'9'
         # Pass
       else
         raise InvalidBigDecimalException.new(str, "Unexpected #{char.inspect} character")
       end
     end
 
+    decimal_end_index = (exponent_index || str.bytesize) - 1
     if decimal_index
+      decimal_count = (decimal_end_index - decimal_index).to_u64
+
       value_str = String.build do |builder|
         # We know this is ASCII, so we can slice by index
         builder.write(str.to_slice[0, decimal_index])
-        builder.write(str.to_slice[decimal_index + 1, str.bytesize - decimal_index - 1])
+        builder.write(str.to_slice[decimal_index + 1, decimal_count])
       end
-
       @value = value_str.to_big_i
-      @scale = (str.bytesize - decimal_index - 1).to_u64
     else
-      @value = str.to_big_i
-      @scale = 0_u64
+      decimal_count = 0_u64
+      @value = str[0..decimal_end_index].to_big_i
+    end
+
+    if exponent_index
+      exponent_postfix = str[exponent_index + 1]
+      case exponent_postfix
+      when '+', '-'
+        exponent_positive = exponent_postfix == '+'
+        exponent = str[(exponent_index + 2)..-1].to_u64
+      else
+        exponent_positive = true
+        exponent = str[(exponent_index + 1)..-1].to_u64
+      end
+
+      @scale = exponent
+      if exponent_positive
+        if @scale < decimal_count
+          @scale = decimal_count - @scale
+        else
+          @scale -= decimal_count
+          @value *= 10.to_big_i ** @scale
```

Comment:
My question still stands: why do we have to apply `@scale` to `@value` here? Isn't the point of scale to represent what you just calculated by hand?

Reply:
Because `@scale` in `BigDecimal` is an `UInt` which represents negative scale, so we can't use it for representing the positive, we need to multiple value by it.

Reply:
oh, OK, fair enough.